Sage-Bionetworks/sage-monorepo
# fix: Unable to build the image of schematic-api

Fixes #1573

## Description

The project `schematic-api` was no longer working. One solution was to update from Python 3.9.2 to Python 3.11.0. I decided to implement this solution as this gave me the opportunity to explore how to enable projects in this monorepo to use different versions of Python if needed.

## Changelog

- Update the project `schematic-api` to use Python 3.11.0
- Fix the task `build-image` of the project `schematic-api` by upgrading to Python 3.11.0
- Add the workspace command `schematic-build-images` to build all the images of the Schematic projects.

## Related PRs

- #1576
- #1582

## Notes

- Installing a Python environment may take more than 1 min (depends on compute resources and internet bandwidth). The time required to perform this action is larger than the time required to download a slightly larger devcontainer image that is preloaded with the environment. For this reason, the Sage devcontainer now accepts multiple Python environments to be included in its image (up to a limit). 

    ```console
    $ time pyenv install 3.11.0
    ...
    real    1m31.234s
    ```

## How to test this PR

- Checkout the branch of this PR
- This branch uses a recently created devcontainer. After checking out the branch, rebuild and reopen the devcontainer.
- Run the commands listed in the section Preview (see below)

## Preview

### Prepare the python environment

```console
nx prepare-python schematic-api
```

### Start the Schematic REST API in development mode

```console
nx serve schematic-api
```

The Swagger UI page is now available on `localhost:7080`

![image](https://github.com/Sage-Bionetworks/sage-monorepo/assets/3056480/bf3cde67-87e7-446f-8ddd-e7d54fd2ab32)

### Build the Docker image

```console
nx build-image schematic-api
```

### Start the Schematic REST API with Docker

```console
nx serve-detach schematic-api
```

The Swagger UI page is now available on `localhost:7080`

### Run the unit tests

```console
nx test schematic-api
```

### Build all the Schematic images

```console
$ schematic-build-images 

    ✔  nx run schematic-api-description:build  [local cache]
    ✔  nx run schematic-api-docs:build  [local cache]
    ✔  nx run schematic-api-docs:build-image (3s)
    ✔  nx run schematic-api:build-image (53s)
```
